src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1771,22 +1866,86 @@ public sealed class One : Tests
                 public override Tests RemoveEvaluation(BoundDagEvaluation e) => e.Equals(Test) ? Tests.True.Instance : (Tests)this;
                 public override string Dump(Func<BoundDagTest, string> dump) => dump(this.Test);
                 public override bool Equals(object? obj) => this == obj || obj is One other && this.Test.Equals(other.Test);
                 public override int GetHashCode() => this.Test.GetHashCode();
+                public override Tests RewriteNestedLengthTests()
+                {
+                    BoundDagTest test = Test;
+                    BoundDagTemp input = test.Input;
+                    if (input.Source is BoundDagPropertyEvaluation { IsLengthOrCount: true } e)
+                    {
+                        if (GetTopLevelLengthTemp(e) is (BoundDagTemp lengthTemp, int offset))
+                        {
+                            switch (test)
+                            {
+                                case BoundDagValueTest { Value.IsBad: false } t:
+                                    {
+                                        Debug.Assert(t.Value.Discriminator == ConstantValueTypeDiscriminator.Int32);
+                                        int lengthValue = t.Value.Int32Value + offset;
+                                        return new One(new BoundDagValueTest(t.Syntax, ConstantValue.Create(lengthValue), lengthTemp));
+                                    }
+                                case BoundDagRelationalTest { Value.IsBad: false } t:
+                                    {
+                                        Debug.Assert(t.Value.Discriminator == ConstantValueTypeDiscriminator.Int32);
+                                        int lengthValue = t.Value.Int32Value + offset;
+                                        if (lengthValue <= 0 && t.Relation == BinaryOperatorKind.GreaterThanOrEqual)
+                                            return True.Instance;
+                                        return new One(new BoundDagRelationalTest(t.Syntax, t.OperatorKind, ConstantValue.Create(lengthValue), lengthTemp));
+                                    }
+                            }
+                        }
+                        else
+                        {
+                            switch (test)
+                            {
+                                case BoundDagRelationalTest { Value.IsBad: false, Relation: BinaryOperatorKind.GreaterThanOrEqual } t:
+                                    {
+                                        Debug.Assert(t.Value.Discriminator == ConstantValueTypeDiscriminator.Int32);
+                                        Debug.Assert(t.OperatorKind == BinaryOperatorKind.IntGreaterThanOrEqual);
+                                        int lengthValue = t.Value.Int32Value;
+                                        if (lengthValue <= 0)
+                                            return True.Instance;
```

Comment:
:memo: Normally this kind of test is omitted in the DAG builder, but we do this here because we're rewriting nested length tests to check the top-level length instead, if we omit the test in the builder, the temp gets removed.